crates/polars-stream/src/nodes/io_sources/ipc.rs

```diff
@@ -0,0 +1,508 @@
+use std::io::Cursor;
+use std::ops::Range;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+
+use polars_core::frame::DataFrame;
+use polars_core::prelude::{AnyValue, ArrowField, ArrowSchema, Column, DataType, IntoColumn};
+use polars_core::scalar::Scalar;
+use polars_core::series::Series;
+use polars_core::utils::arrow::array::Array;
+use polars_core::utils::arrow::io::ipc::read::{read_file_metadata, FileMetadata, FileReader};
+use polars_error::PolarsResult;
+use polars_expr::prelude::PhysicalExpr;
+use polars_expr::state::ExecutionState;
+use polars_io::cloud::CloudOptions;
+use polars_io::ipc::IpcScanOptions;
+use polars_io::utils::{apply_projection, columns_to_projection};
+use polars_io::RowIndex;
+use polars_plan::plans::hive::HivePartitions;
+use polars_plan::plans::{FileInfo, ScanSources};
+use polars_plan::prelude::FileScanOptions;
+use polars_utils::mmap::MemSlice;
+use polars_utils::pl_str::PlSmallStr;
+use polars_utils::IdxSize;
+
+use crate::async_primitives::distributor_channel::distributor_channel;
+use crate::async_primitives::wait_group::WaitGroup;
+use crate::morsel::{get_ideal_morsel_size, SourceToken};
+use crate::nodes::{
+    ComputeNode, JoinHandle, Morsel, MorselSeq, PortState, TaskPriority, TaskScope,
+};
+use crate::pipe::{RecvPort, SendPort};
+use crate::DEFAULT_DISTRIBUTOR_BUFFER_SIZE;
+
+pub struct IpcSourceNode {
+    sources: ScanSources,
+
+    row_index: Option<RowIndex>,
+    projection: Option<Vec<usize>>,
+    slice: Option<(i64, usize)>,
+
+    rechunk: bool,
+    include_file_paths: Option<PlSmallStr>,
+
+    projected_schema: Arc<ArrowSchema>,
+
+    /// Can the SendPort be closed?
+    is_finished: AtomicBool,
```

Comment:
If you separate out the static config `IpcSourceNode` as an `IpcSourceConfig` and pass that instead of `slf` into the other workers, you can keep mutable access to `self` and this doesn't have to be atomic.